lib/kredis/types/set.rb

```diff
@@ -4,7 +4,8 @@ class Kredis::Types::Set < Kredis::Types::Proxying
   attr_accessor :typed
 
   def members
-    strings_to_types(smembers || [], typed).sort
+    value = exists? ? smembers : initialize_with_default || []
```

Comment:
This adds an additional redis command call on every call to #members. Why wouldn't `smembers || initialize_with_default || []` work?

Reply:
From my testing, when a _set_ has not been created, redis returns the following:

- `#exists?` => false
- `#smembers` => [] # empty array

I assumed we should not use just an empty array to decide if the code should apply the default value.

Reply:
According to this answer on StackOverflow,

> When a set is empty, it is removed from the namespace.

see: https://stackoverflow.com/questions/13817865/redis-nil-or-empty-list-or-set#13820112

So, maybe we can assume with:

```ruby
strings_to_types(smembers.presence || initialize_with_default || [], typed).sort
```

Reply:
That looks right to me 👍

Reply:
Same question might go for list?

```diff
  def elements
-   value = exists? ? lrange(0, -1) : initialize_with_default || []
+   value = lrange(0, -1).presence || initialize_with_default || []
    strings_to_types(value, typed)
  end
```

Reply:
Yes, same issue. Each redis command carries a cost over the network. Need to make sure we don't regress on that.

Reply:
For hash and counter, this is going to be a challenge then.  

both of them implement a `before_methods` which checks if key exists before performing functions.

Reply:
I think we have to find another way, then. We can't tax every hash and counter usage with an extra redis command just to support default.

Reply:
We can solve this for counter, but not for Hash, List, Set, etc.  We can check if default is defined before checking `#exists?`, but I do not know of another way and open to suggestions.

Reply:
Maybe @Jeremy has an idea.

Reply:
Did a fairly significant refactor.  

- Used the same pattern throughout of `#initialize_with_default` defined in the base class.  `#initialize_with_default` calls `#set_default`, which is overridden in many of the child classes.
- Created a custom `#callnx` method which executes a redis lua script.  The lua script _calls_ a dynamic method with values if the redis key does not exist.  (Note: the script uses `redis#eval`, so additional care was taken to validate formatting of the method call.)  Using this custom redis method allows the code to run methods like hset and rpush only if key does not exist, and do this logic without adding redis command calls as it can be executed in a `multi` block
- Updated the custom redis `#multi` method to allow nested multi blocks (when one method using a multi block calls another method which also uses a multi block)

@dhh @jeremy - let me know if this a direction you want to take.

Reply:
FYI, I tested performance with the following benchmarks


```ruby
require 'benchmark/ips'

proxy = Kredis.proxy 'mykey'
proxy.set 'myvalue'

Benchmark.ips(7) do |x|
  x.report("get only") do
    proxy = Kredis.proxy 'mykey'
    proxy.get
  end
  x.report("exists?") do
    proxy = Kredis.proxy 'mykey'
    proxy.set 'myvalue' unless proxy.exists?
    proxy.get
  end
  x.report("nx: true") do
    proxy = Kredis.proxy 'mykey'
    proxy.set 'myvalue', nx: true
    proxy.get
  end
  x.report("callnx") do
    proxy = Kredis.proxy 'mykey'
    proxy.callnx(:set, 'myvalue')
    proxy.get
  end
  x.report("multi > nx: true") do
    proxy = Kredis.proxy 'mykey'
    proxy.multi do
      proxy.set 'myvalue', nx: true
      proxy.get
    end
  end
  x.report("multi > callnx") do
    proxy = Kredis.proxy 'mykey'
    proxy.multi do
      proxy.callnx(:set, 'myvalue')
      proxy.get
    end
  end
  x.compare!
end
```

The results show:

```
Warming up --------------------------------------
            get only     1.064k i/100ms
             exists?   731.000  i/100ms
            nx: true   644.000  i/100ms
              callnx   621.000  i/100ms
    multi > nx: true   702.000  i/100ms
      multi > callnx   664.000  i/100ms
Calculating -------------------------------------
            get only     10.617k (± 6.7%) i/s -     74.480k in   7.053390s
             exists?      7.322k (± 2.9%) i/s -     51.901k in   7.094167s
            nx: true      7.030k (± 5.7%) i/s -     49.588k in   7.080403s
              callnx      6.810k (± 7.1%) i/s -     47.817k in   7.059580s
    multi > nx: true      6.891k (± 4.7%) i/s -     48.438k in   7.047004s
      multi > callnx      6.608k (± 4.2%) i/s -     46.480k in   7.046974s

Comparison:
            get only:    10617.3 i/s
             exists?:     7322.1 i/s - 1.45x  (± 0.00) slower
            nx: true:     7030.3 i/s - 1.51x  (± 0.00) slower
    multi > nx: true:     6891.2 i/s - 1.54x  (± 0.00) slower
              callnx:     6810.3 i/s - 1.56x  (± 0.00) slower
      multi > callnx:     6607.6 i/s - 1.61x  (± 0.00) slower
```

#### Observations:
- As expected, that get only is the fastest
- As expected, defining a default value adds cost in checking the existence and setting if not exist
- Perhaps something is wrong with the test setup, but in these scenarios using `multi` adds additional overhead vs not using `multi`.  I expected the opposite.  So, I setup a 2nd test around the usage of `multi`.

```ruby
require 'benchmark/ips'

proxy = Kredis.proxy 'mykey'
proxy.set 'myvalue'

n = 1
Benchmark.ips(7) do |x|
  x.report("1 cmd") do
    proxy = Kredis.proxy 'mykey'
    proxy.get
  end
  x.report("2 cmds") do
        proxy = Kredis.proxy 'mykey'
        proxy.set 'myvalue'
        proxy.get
      end
      x.report("3 cmds") do
        proxy = Kredis.proxy 'mykey'
        proxy.set 'myvalue'
        proxy.set 'myvalue'
        proxy.get
      end
      x.report("multi > 1 cmd") do
        proxy = Kredis.proxy 'mykey'
        proxy.multi do
          proxy.get
        end
      end
      x.report("multi > 2 cmds") do
        proxy = Kredis.proxy 'mykey'
        proxy.multi do
          proxy.set 'myvalue'
          proxy.get
        end
      end
      x.report("multi > 3 cmds") do
        proxy = Kredis.proxy 'mykey'
        proxy.multi do
          proxy.set 'myvalue'
          proxy.set 'myvalue'
          proxy.get
        end
      end
      x.compare!
    end
```

```
Warming up --------------------------------------
               1 cmd     1.059k i/100ms
              2 cmds   709.000  i/100ms
              3 cmds   552.000  i/100ms
       multi > 1 cmd   797.000  i/100ms
      multi > 2 cmds   730.000  i/100ms
      multi > 3 cmds   610.000  i/100ms
Calculating -------------------------------------
               1 cmd     10.290k (± 6.8%) i/s -     72.012k in   7.034624s
              2 cmds      7.160k (± 3.1%) i/s -     50.339k in   7.037131s
              3 cmds      5.289k (± 6.5%) i/s -     36.984k in   7.026510s
       multi > 1 cmd      7.937k (± 5.1%) i/s -     55.790k in   7.049686s
      multi > 2 cmds      6.976k (± 3.7%) i/s -     48.910k in   7.021257s
      multi > 3 cmds      6.317k (± 5.0%) i/s -     44.530k in   7.069582s

Comparison:
               1 cmd:    10289.5 i/s
       multi > 1 cmd:     7937.3 i/s - 1.30x  (± 0.00) slower
              2 cmds:     7160.2 i/s - 1.44x  (± 0.00) slower
      multi > 2 cmds:     6975.8 i/s - 1.48x  (± 0.00) slower
      multi > 3 cmds:     6316.9 i/s - 1.63x  (± 0.00) slower
              3 cmds:     5289.0 i/s - 1.95x  (± 0.00) slower
```

#### Observations:
- Using multi adds overhead
- Using multi is faster once you reach 3+ commands.  Using only 1 or 2 is slower
- In the latest updated version there's lots of usage of `multi` (which I assumed would be faster).  After seeing these benchmarks, it's going to slightly slow down the current performance.  For example:

```diff
def value
-     value_after_casting = string_to_type(get, typed)

-     if value_after_casting.nil?
-       default
-     else
-       value_after_casting
-     end
+    get_value = multi do
+         set type_to_string(value, typed), ex: expires_in, nx: true if default_value.present?
+         get
+    end[-1]
+    string_to_type(get_value, typed)
end
```

In the above example, even if default is nil the code will run the `get` inside of `multi`, which goes from 10289.5 i/s to 7937.3 i/s.

Based on these benchmarks, I suggest either going back to using `#exists?` (if code defines default).

Reply:
I have updated the code to be more efficient in that it only uses `multi` in the event that a default value has been defined, otherwise it reverts back to the prior behavior.

At this point, the code is ready for another review.